### DISCUSSION TITLE
Can't use LoginLinkNotification from a console SF app

### DISCUSSION THREAD

**Discussion by klamorski** [symfony/symfony#43279] [Q&A] [answered]
I try to use LoginLinkNotification from a console application to send invitation e-mails with custom login links to our all users. It is not possible. 
The app generates the following error message: "In FirewallAwareTrait.php line 32: Cannot determine the correct Symfony\Bundle\SecurityBundle\LoginLink\LoginLinkHandler to use: there is no active Request and so, the firewall cannot be determined. Try using a specific Symfony\Bundle\SecurityBundle\LoginLink\LoginLinkHandler service."

If I understand the problem correctly, the Request context is missing in the console application which is a source of the problem.

May I overcome these problems somehow? The Symfony 5.3.9 is used in this project.

The code causing this issue below:
```php
namespace App\Command;

use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\Style\SymfonyStyle;


use App\Entity\User;
use Doctrine\ORM\EntityManagerInterface;

use Symfony\Component\Notifier\NotifierInterface;
use Symfony\Component\Notifier\Recipient\Recipient;
use Symfony\Component\Security\Http\LoginLink\LoginLinkNotification;
use Symfony\Component\Security\Http\LoginLink\LoginLinkHandlerInterface;
use Symfony\Component\Routing\RouterInterface;
use Symfony\Component\HttpFoundation\Request;

class SendMailInvitationsCommand extends Command
{
    protected static $defaultName = 'app:send-mail-invitations';

    
    private $entityManager;
    private $notifier;
    private $loginLinkHandler;

    public function __construct(RouterInterface $router, EntityManagerInterface $entityManager, NotifierInterface $notifier, LoginLinkHandlerInterface $loginLinkHandler)
    {
        $this->entityManager = $entityManager;
        $this->notifier = $notifier;
        $this->loginLinkHandler = $loginLinkHandler;
        parent::__construct();
    }

    protected function configure(): void
    {
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $users = $this->entityManager->getRepository(User::class)->findAll();
        foreach ($users as $user) {
            $output->writeln($user->getEmail());
            $loginLinkDetails = $this->loginLinkHandler->createLoginLink($user);
            $notification = new LoginLinkNotification($loginLinkDetails);
            $recipient = new Recipient($user->getEmail());
            $this->notifier->send($notification, $recipient);
        }
        return Command::SUCCESS;
    }
}

```

**Comment by wouterj** [comment, marked as answer]
The issue is with the `LoginLinkHandlerInterface $loginLinkHandler` dependency in your constructor. This dependency autowiring requires a request (e.g. if you have 2 firewalls in your application with a login link authentication for both of them, the request is used to determine which of the 2 login link handlers you want).

Each login link handler service uses the following name: `security.authenticator.login_link_handler.<FIREWALL NAME>`. You can use this to configure the dependency for your service to skip autowiring (and thus remove the dependency on request). E.g. if your firewall is called "main", you need to use something like:

```
# config/services.yaml
services:
    # ...

    App\Command\SendMailInvitationsCommand:
        arguments:
            $loginLinkHandler: '@security.authenticator.login_link_handler.main'
```


Note that because you're generating URLs in the console, you need to configure the "default uri" for the Symfony router: https://symfony.com/doc/current/routing.html#generating-urls-in-commands

**Reply by klamorski** [reply]
Works like a charm. Thank you very much for the time you spent explaining the problem to me.

**Comment by pieterdt** [comment]
I have both  the servces.yaml configuration in place and the default uri. But still getting the same error. Could this be because I'm not using a console command, but a process run by a worker?
